File: services/assessment-service/src/main/java/com/visor/school/assessment/service/GradebookService.java
package com.visor.school.assessment.service;

import com.visor.school.assessment.model.Assessment;
import com.visor.school.assessment.model.Grade;
import com.visor.school.assessment.repository.AssessmentRepository;
import com.visor.school.assessment.repository.GradeRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Gradebook service for viewing class and student gradebooks
 */
@Service
@Transactional(readOnly = true)
public class GradebookService {
    
    private final AssessmentRepository assessmentRepository;
    private final GradeRepository gradeRepository;
    private final GradeCalculator gradeCalculator;

    public GradebookService(AssessmentRepository assessmentRepository,
                           GradeRepository gradeRepository,
                           GradeCalculator gradeCalculator) {
        this.assessmentRepository = assessmentRepository;
        this.gradeRepository = gradeRepository;
        this.gradeCalculator = gradeCalculator;
    }

    /**
     * Get class gradebook with all assessments and student grades
     */
    public ClassGradebook getClassGradebook(UUID classId) {
        List<Assessment> assessments = assessmentRepository.findByClassId(classId);
        List<UUID> assessmentIds = assessments.stream()
            .map(Assessment::getId)
            .collect(Collectors.toList());

        Map<UUID, List<Grade>> studentGrades = new HashMap<>();

        // Get all grades for assessments in this class
        for (UUID assessmentId : assessmentIds) {
            List<Grade> grades = gradeRepository.findByAssessmentId(assessmentId);
            for (Grade grade : grades) {
                studentGrades.computeIfAbsent(grade.getStudentId(), k -> new ArrayList<>()).add(grade);
            }
        }

        // Calculate averages for each student
        Map<UUID, BigDecimal> studentAverages = studentGrades.entrySet().stream()
            .collect(Collectors.toMap(
                Map.Entry::getKey,
                entry -> gradeCalculator.calculateAverage(entry.getValue())
            ));

        return new ClassGradebook(classId, assessments, studentGrades, studentAverages);
    }

    /**
     * Get student gradebook with all assessments and grades
     */
    public StudentGradebook getStudentGradebook(UUID studentId) {
        List<Grade> grades = gradeRepository.findByStudentId(studentId);
        BigDecimal average = gradeCalculator.calculateAverage(grades);

        return new StudentGradebook(studentId, grades, average);
    }
}

File: services/academic-assessment-service/src/main/kotlin/com/visor/school/assessment/service/GradebookService.kt
package com.visor.school.assessment.service

import com.visor.school.assessment.model.Assessment
import com.visor.school.assessment.model.Grade
import com.visor.school.assessment.repository.AssessmentRepository
import com.visor.school.assessment.repository.GradeRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.math.BigDecimal
import java.util.UUID

/**
 * Gradebook service for viewing class and student gradebooks
 */
@Service
@Transactional(readOnly = true)
class GradebookService(
    private val assessmentRepository: AssessmentRepository,
    private val gradeRepository: GradeRepository,
    private val gradeCalculator: GradeCalculator
) {

    /**
     * Get class gradebook with all assessments and student grades
     */
    fun getClassGradebook(classId: UUID): ClassGradebook {
        val assessments = assessmentRepository.findByClassId(classId)
        val assessmentIds = assessments.map { it.id }

        val studentGrades = mutableMapOf<UUID, MutableList<Grade>>()

        // Get all grades for assessments in this class
        assessmentIds.forEach { assessmentId ->
            val grades = gradeRepository.findByAssessmentId(assessmentId)
            grades.forEach { grade ->
                studentGrades.getOrPut(grade.studentId) { mutableListOf() }.add(grade)
            }
        }

        // Calculate averages for each student
        val studentAverages = studentGrades.mapValues { (_, grades) ->
            gradeCalculator.calculateAverage(grades)
        }

        return ClassGradebook(
            classId = classId,
            assessments = assessments,
            studentGrades = studentGrades,
            studentAverages = studentAverages
        )
    }

    /**
     * Get student gradebook with all assessments and grades
     */
    fun getStudentGradebook(studentId: UUID): StudentGradebook {
        val grades = gradeRepository.findByStudentId(studentId)
        val assessmentIds = grades.map { it.assessmentId }.distinct()

        val assessments = assessmentIds.mapNotNull { assessmentId ->
            // Note: In a real implementation, you might want to fetch assessments from the Academic Service
            // For now, we'll use the assessment IDs from grades
            null // Placeholder - would need to fetch from Academic Service
        }

        val average = gradeCalculator.calculateAverage(grades)

        return StudentGradebook(
            studentId = studentId,
            grades = grades,
            average = average
        )
    }
}

data class ClassGradebook(
    val classId: UUID,
    val assessments: List<Assessment>,
    val studentGrades: Map<UUID, List<Grade>>,
    val studentAverages: Map<UUID, BigDecimal>
)

data class StudentGradebook(
    val studentId: UUID,
    val grades: List<Grade>,
    val average: BigDecimal
)

